src/sinks/prometheus/remote_write.rs

```diff
@@ -0,0 +1,269 @@
+        let tls_settings = TlsSettings::from_options(&self.tls)?;
+        let batch = BatchSettings::default()
+            .events(1_000)
+            .timeout(1)
+            .parse_config(self.batch)?;
+        let request = self.request.unwrap_with(&REQUEST_DEFAULTS);
+        let buckets = self.buckets.clone();
+        let quantiles = self.quantiles.clone();
+
+        let client = HttpClient::new(tls_settings)?;
+        let healthcheck = healthcheck(endpoint.clone(), client.clone()).boxed();
+        let service = RemoteWriteService {
+            endpoint,
+            client,
+            buckets,
+            quantiles,
+        };
+        let sink = request
+            .batch_sink(
+                HttpRetryLogic,
+                service,
+                MetricBuffer::new(batch.size),
+                batch.timeout,
+                cx.acker(),
+            )
+            .sink_map_err(|error| error!(message = "Prometheus remote_write sink error.", %error));
+
+        Ok((
+            sinks::VectorSink::Futures01Sink(Box::new(sink)),
+            healthcheck,
+        ))
+    }
+
+    fn input_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "prometheus_remote_write"
+    }
+}
+
+async fn healthcheck(endpoint: Uri, mut client: HttpClient) -> crate::Result<()> {
+    let request = http::Request::get(endpoint)
+        .body(hyper::Body::empty())
+        .unwrap();
+
+    let response = client.send(request).await?;
+
+    match response.status() {
+        http::StatusCode::OK => Ok(()),
+        other => Err(sinks::HealthcheckError::UnexpectedStatus { status: other }.into()),
+    }
+}
+
+#[derive(Clone)]
+struct RemoteWriteService {
+    endpoint: Uri,
+    client: HttpClient,
+    buckets: Vec<f64>,
+    quantiles: Vec<f64>,
+}
+
+impl RemoteWriteService {
+    fn encode_events(&self, metrics: Vec<Metric>) -> crate::Result<Bytes> {
+        let mut time_series = collector::TimeSeries::new();
+        for metric in metrics {
+            time_series.encode_metric(None, &self.buckets, &self.quantiles, false, &metric);
+        }
+        let timeseries = time_series.finish();
+
+        let request = proto::WriteRequest { timeseries };
+        let mut out = BytesMut::with_capacity(request.encoded_len());
+        request
+            .encode(&mut out)
+            .map(move |_| out.freeze())
+            .map_err(Into::into)
+    }
+}
+
+impl tower::Service<Vec<Metric>> for RemoteWriteService {
+    type Response = http::Response<Bytes>;
+    type Error = crate::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _task: &mut task::Context<'_>) -> task::Poll<Result<(), Self::Error>> {
+        task::Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, events: Vec<Metric>) -> Self::Future {
+        let body = self.encode_events(events).unwrap();
```

Comment:
Is this `unwrap` because we're just writing to memory? Could be nice to note that with `expect` if so.

Reply:
The only way for encoding to fail is if the buffer has insufficient capacity. We are encoding to `BytesBuf` which grows as needed, so this is impossible. Still, it would probably be better to strip the error within `encode_events` since we can't do anything about it.